aliasmgr.go

```diff
@@ -0,0 +1,446 @@
+	a.RLock()
+	defer a.RUnlock()
+
+	aliasSet, ok := a.baseToSet[base]
+	if ok {
+		// Copy the found alias slice.
+		setCopy := make([]lnwire.ShortChannelID, 0, len(aliasSet))
+		setCopy = append(setCopy, aliasSet...)
+		return setCopy
+	}
+
+	return nil
+}
+
+// findBaseSCID finds the base SCID for a given alias. This is used in the
+// gossiper to find the correct SCID to lookup in the graph database.
+func (a *aliasMgr) findBaseSCID(
+	alias lnwire.ShortChannelID) (lnwire.ShortChannelID, error) {
+
+	a.RLock()
+	defer a.RUnlock()
+
+	base, ok := a.aliasToBase[alias]
+	if ok {
+		return base, nil
+	}
+
+	return lnwire.ShortChannelID{}, errNoBase
+}
+
+// deleteSixConfs removes a mapping for the gossiper once six confirmations
+// have been reached and the channel is public. At this point, only the
+// confirmed SCID should be used.
+func (a *aliasMgr) deleteSixConfs(baseScid lnwire.ShortChannelID) error {
+	a.Lock()
+	defer a.Unlock()
+
+	err := kvdb.Update(a.backend, func(tx kvdb.RwTx) error {
+		baseConfBucket, err := tx.CreateTopLevelBucket(confirmedBucket)
+		if err != nil {
+			return err
+		}
+
+		var baseBytes [8]byte
+		byteOrder.PutUint64(baseBytes[:], baseScid.ToUint64())
+		return baseConfBucket.Put(baseBytes[:], []byte{})
+	}, func() {})
+	if err != nil {
+		return err
+	}
+
+	// Now that the database state has been updated, we'll delete all of
+	// the aliasToBase mappings for this SCID.
+	for alias, base := range a.aliasToBase {
+		if base.ToUint64() == baseScid.ToUint64() {
+			delete(a.aliasToBase, alias)
+		}
+	}
+
+	return nil
+}
+
+// putPeerAlias stores the peer's alias SCID once we learn of it in the
+// funding_locked message.
+func (a *aliasMgr) putPeerAlias(chanID lnwire.ChannelID,
+	alias lnwire.ShortChannelID) error {
+
+	return kvdb.Update(a.backend, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(invoiceAliasBucket)
+		if err != nil {
+			return err
+		}
+
+		var scratch [8]byte
+		byteOrder.PutUint64(scratch[:], alias.ToUint64())
+		return bucket.Put(chanID[:], scratch[:])
+	}, func() {})
+}
+
+// getPeerAlias retrieves a peer's alias SCID by the channel's ChanID.
+func (a *aliasMgr) getPeerAlias(chanID lnwire.ChannelID) (
+	lnwire.ShortChannelID, error) {
+
+	var alias lnwire.ShortChannelID
+
+	err := kvdb.Update(a.backend, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(invoiceAliasBucket)
+		if err != nil {
+			return err
+		}
+
+		aliasBytes := bucket.Get(chanID[:])
+		if aliasBytes == nil {
+			return nil
+		}
+
+		alias = lnwire.NewShortChanIDFromInt(
+			byteOrder.Uint64(aliasBytes),
+		)
+		return nil
+	}, func() {})
+
+	if alias == hop.Source {
+		return alias, errNoPeerAlias
+	}
+
+	return alias, err
+}
+
+// requestAlias returns a new ALIAS ShortChannelID to the caller by allocating
+// the next un-allocated ShortChannelID. The starting ShortChannelID is
+// 10000:0:0 and the ending ShortChannelID is 262143:16777215:65535. This gives
```

Comment:
Don't want it to conflict with public lightning channels as it would have the potential to mess up forwarding if our counter-party receives alias `{550222:2:2}` and they actually have that same lightning channel, where do they forward to?

Reply:
Editing the comment above:

The spec requires that we don't give out an alias A to one peer and then give the same alias A to another peer. What it does not require is that if we give out an alias A, that A does not conflict with any _existing_ short channel id's we may have. Maybe we should bring this up? Because in this case, our counter-party will issue an invoice with alias A, the sender sends us an onion with the next hop as alias A, but we have _two_ channels with alias A (one option-scid-alias channel, one regular channel w/ a regular SCID created before the spec change)